### PR TITLE
Limit Autofailure condition

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 

1. Default the `spec.tikv.maxfailureOver` and `spec.tidb.maxfailureOver` to 3 when they are no defined
2. Change the action that when `maxfailureOver` is zero, operator would not failover instead of creating limitless instances.
3. Still Checking the `maxfailureOver` whether it is nil in the failover check.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### Does this PR introduce a user-facing change?:

```release-note
1. Default the `spec.tikv.maxfailureOver` and `spec.tidb.maxfailureOver` to 3 when they are no defined
2. Change the action that when `maxfailureOver` is zero, operator would not failover instead of creating limitless instances.
```

### CODE REVIEW

File: pkg/manager/member/tidb_member_manager.go [hunk 1]
```diff
@@ -198,7 +198,7 @@ func (tmm *tidbMemberManager) syncTiDBStatefulSetForTidbCluster(tc *v1alpha1.Tid
 		}
 	}
 
-	if tmm.autoFailover {
+	if tmm.autoFailover && tc.Spec.TiDB.MaxFailoverCount != nil {
```
Comment (cofyc) [inline pkg/manager/member/tidb_member_manager.go]: what's the default value?

Comment (cofyc) [reply pkg/manager/member/tidb_member_manager.go]: I seem that we don't have a default value, so the user must specify this field to enable the auto-failover feature? @aylei @weekface what do you think?

Comment (Yisaer) [reply pkg/manager/member/tidb_member_manager.go]: Yes, it is ok to me to set the default value to ensure the `MaxFailoverCount` is not nil`.  But I still think it is necessary to check whether the `MaxFailoverCount` is nil here.

File: pkg/manager/member/tidb_failover.go [hunk 2]
```diff
@@ -45,22 +45,22 @@ func (tf *tidbFailover) Failover(tc *v1alpha1.TidbCluster) error {
...
 		}
-	}
-	for _, tidbMember := range tc.Status.TiDB.Members {
-		_, exist := tc.Status.TiDB.FailureMembers[tidbMember.Name]
-		deadline := tidbMember.LastTransitionTime.Add(tf.tidbFailoverPeriod)
-		if !tidbMember.Health && time.Now().After(deadline) && !exist {
-			tc.Status.TiDB.FailureMembers[tidbMember.Name] = v1alpha1.TiDBFailureMember{
-				PodName:   tidbMember.Name,
-				CreatedAt: metav1.Now(),
+		for _, tidbMember := range tc.Status.TiDB.Members {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidb_failover.go]: Why move this for loop in the control of failover? We may still need to update the status for the failure members even if the failover is disabled.

File: pkg/manager/member/tidb_member_manager.go (hunk 1 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tidb_member_manager.go]: The `tc.Spec.TiDB.MaxFailoverCount` will never be nil after adding the defaulting function, right?

File: pkg/manager/member/tikv_failover.go [hunk 3]
```diff
@@ -51,18 +51,17 @@ func (tf *tikvFailover) Failover(tc *v1alpha1.TidbCluster) error {
...
 					klog.Warningf("%s/%s failure stores count reached the limit: %d", ns, tcName, tc.Spec.TiKV.MaxFailoverCount)
 					return nil
 				}
-			}
-
-			tc.Status.TiKV.FailureStores[storeID] = v1alpha1.TiKVFailureStore{
-				PodName:   podName,
-				StoreID:   store.ID,
-				CreatedAt: metav1.Now(),
+				tc.Status.TiKV.FailureStores[storeID] = v1alpha1.TiKVFailureStore{
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_failover.go]: Suggest not moving this block under the control of failover count.

File: pkg/manager/member/tikv_member_manager.go [hunk 4]
```diff
@@ -223,7 +223,7 @@ func (tkmm *tikvMemberManager) syncStatefulSetForTidbCluster(tc *v1alpha1.TidbCl
 		return err
 	}
 
-	if tkmm.autoFailover {
+	if tkmm.autoFailover && tc.Spec.TiKV.MaxFailoverCount != nil {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_member_manager.go]: The pointer will never be nil after defaulting.

File: pkg/controller/tidbcluster/tidb_cluster_control.go [hunk 5]
```diff
@@ -114,6 +115,8 @@ func (tcc *defaultTidbClusterControl) validate(tc *v1alpha1.TidbCluster) bool {
 }
 
 func (tcc *defaultTidbClusterControl) updateTidbCluster(tc *v1alpha1.TidbCluster) error {
+	// default the tidbcluster
+	defaulting.SetTidbClusterDefault(tc)
```
Comment (cofyc) [inline pkg/controller/tidbcluster/tidb_cluster_control.go]: move it to this function, we can have our controller-side validation and defaulting at the same place

https://github.com/pingcap/tidb-operator/blob/52491834b6e0371a4ae9bd862401f4d398d27e61/pkg/controller/tidbcluster/tidb_cluster_control.go#L84-L87

File: pkg/manager/member/tidb_failover.go (hunk 2 above)
Comment (cofyc) [reply pkg/manager/member/tidb_failover.go]: yes, the `for _, tidbMember := range tc.Status.TiDB.Members {` loop is part of the status syncing. the user can choose not to failover when a member fails, but the controller should record failure members in the status always.

File: pkg/manager/member/tikv_member_manager.go (hunk 4 above)
Comment (Yisaer) [reply pkg/manager/member/tikv_member_manager.go]: I think the defaulting and checking are 2 different things.

File: pkg/manager/member/tidb_failover.go (hunk 2 above)
Comment (Yisaer) [reply pkg/manager/member/tidb_failover.go]: I think there is no updating status here both in the previous or in the current.

File: pkg/controller/tidbcluster/tidb_cluster_control.go (hunk 5 above)
Comment (Yisaer) [reply pkg/controller/tidbcluster/tidb_cluster_control.go]: updated.

**Comment by cofyc** [comment]
https://github.com/pingcap/tidb-operator/blob/52491834b6e0371a4ae9bd862401f4d398d27e61/pkg/apis/pingcap/v1alpha1/types.go#L260-L264

https://github.com/pingcap/tidb-operator/blob/52491834b6e0371a4ae9bd862401f4d398d27e61/pkg/apis/pingcap/v1alpha1/types.go#L307-L311

btw, comments of these two fields must be updated

File: pkg/manager/member/tikv_failover.go (hunk 3 above)
Comment (cofyc) [reply pkg/manager/member/tikv_failover.go]: don't record failure tikv store if auto-failover is disabled? I guess this not expected.

Comment (cofyc) [reply pkg/manager/member/tikv_failover.go]: oh, I see that this is the same as Failmember in tidb status...my mistake.

**Pull request merged**